### PR TITLE
Minimal json api

### PR DESCRIPTION
**Pull request by martinhoefling**
Implementation of a minimal set of messages (query and getLogin + Responses). Integration tests for use and edge cases are implemented. Let me know if something is missing. Further, I rebased on two commits of @dominikschulz to make the tests run on OSX - so merging is only possible after his commits are reviewed and merged. Please comment and discuss the implementation.

### CODE REVIEW

File: action/jsonapi.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
+package action
+
+import (
+	"bufio"
```
Comment (dominikschulz) [inline action/jsonapi.go]: Did you format this file with gofmt / goimports? The import order looks random.

File: action/jsonapi.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	Password string `json:"password"`
+}
+
+type errorResponse struct {
+	Error string `json:"error"`
+}
+
+// JSONAPI reads a json message on stdin and responds on stdout
+func (s *Action) JSONAPI(ctx context.Context, c *cli.Context) error {
+	err := readAndRespond(ctx, s)
```
Comment (dominikschulz) [inline action/jsonapi.go]: I'd prefer the idiom `if err := ...; err != nil {`

File: action/jsonapi.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		var response errorResponse
+		response.Error = err.Error()
+
+		err = sendSerializedJSONMessage(response)
```
Comment (dominikschulz) [inline action/jsonapi.go]: `if err := ...; err != nil {`, please

File: action/jsonapi.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+func readAndRespond(ctx context.Context, s *Action) error {
+	message, err := readMessage()
+	if message != nil {
+		err = respondMessage(ctx, s, message)
```
Comment (dominikschulz) [inline action/jsonapi.go]: `if err := ...; err != nil {`, please

File: action/jsonapi.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func readAndRespond(ctx context.Context, s *Action) error {
+	message, err := readMessage()
+	if message != nil {
```
Comment (dominikschulz) [inline action/jsonapi.go]: I'd prefer to handle the error case in the block and the success case on the upper level, e.g. `if err != nil || message == nil { <error handling>  } <success handling>`

File: action/jsonapi.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return nil, eofReturn(err)
+	}
+	return msgBytes, nil
+}
+
+func getMessageLength(msg []byte) (int, error) {
+	var length uint32
+	buf := bytes.NewBuffer(msg)
+	err := binary.Read(buf, binary.LittleEndian, &length)
```
Comment (dominikschulz) [inline action/jsonapi.go]: `if err := ...; err != nil {`, please

File: action/jsonapi.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+func eofReturn(err error) error {
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+func respondMessage(ctx context.Context, s *Action, msgBytes []byte) error {
+	var message messageType
+	err := json.Unmarshal(msgBytes, &message)
```
Comment (dominikschulz) [inline action/jsonapi.go]: `if err := ...; err != nil {`, please

File: action/jsonapi.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	return sendSerializedJSONMessage(choices)
+}
+
+func respondGetLogin(ctx context.Context, s *Action, msgBytes []byte) error {
+	var message getLoginMessage
+	var response loginResponse
+
+	err := json.Unmarshal(msgBytes, &message)
```
Comment (dominikschulz) [inline action/jsonapi.go]: `if err := ...; err != nil {`, please

File: action/jsonapi.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return err
+	}
+
+	secret, err := s.Store.Get(ctx, message.Entry)
+
```
Comment (dominikschulz) [inline action/jsonapi.go]: please remove this blank line, the error handling should visually stick to it's (possible) cause

File: action/jsonapi.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+		return err
+	}
+
+	response.Username, err = getUsername(ctx, s, message.Entry)
+	if err != nil {
+		return err
+	}
+	response.Password = secret.Password()
+
+	err = sendSerializedJSONMessage(response)
```
Comment (dominikschulz) [inline action/jsonapi.go]: just `return sendSerializedJSONMessage(response)`

File: action/jsonapi.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+func sendSerializedJSONMessage(message interface{}) error {
+	var msgBuf bytes.Buffer
+
+	serialized, err := json.Marshal(message)
+	if err != nil {
+		return err
+	}
+
+	err = writeMessageLength(serialized)
```
Comment (dominikschulz) [inline action/jsonapi.go]: `if err := ...; err != nil {`, please

**Comment by martinhoefling** [comment]
Concerning the reference documentation. Is it ok to create a docs/jsonapi.md and reference it from the top level README.md? I would put specifics in jsonapi.md. Once the browser plugin is stabilized and published, I would put some references in the jsonapi documentation of README.md.

File: action/jsonapi.go (hunk 1 above)
Comment (martinhoefling) [reply action/jsonapi.go]: Yeah, I missed that. gofmt and goimports applied to the two new files.

**Comment by dominikschulz** [comment]
Yes, that would be very good.

**Comment by metalmatze** [comment]
We have merged the PRs that this is depending on. If you could be so kind to rebase your PR. Thanks!

**Comment by dominikschulz** [comment]
@martinhoefling Could you please rebase?

**Comment by martinhoefling** [comment]
@dominikschulz [x] done

**Comment by martinhoefling** [comment]
Documentation for JSON API added
